### PR TITLE
[dynamo] Support __next__ on UserDefinedObjectVariable

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122565

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -362,6 +362,56 @@ def forward(
...
+            raise AssertionError()
+
+    def __iter__(self):
+        return self._iter_ex(True)
+
+    def _iter_ex(self, resolve: bool) -> Iterator[Any]:
+        try:
+            return ListConfig.ListIterator(self, resolve)
+        except Exception as e:
+            raise AssertionError()
```
Comment (Skylion007) [inline test/dynamo/test_repros.py:405]: From e or come None here?

Comment (jansel) [reply test/dynamo/test_repros.py:405]: This code is unreachable, it shouldn't matter.  It is just about generating the try/except bytecodes for the testcase.

**Review by yanboliang — APPROVED** [review]

**Pull request closed without merging**